File: leetcode/src/00967_numbers_with_same_consecutive_differences/solution.go
package leetcode

func numsSameConsecDiff(n int, k int) []int {
	if n == 1 {
		return []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
	}
	var res []int
	for i := 1; i < 10; i++ {
		dfs(n-1, k, i, &res)
	}
	return res
}

func dfs(n int, k int, num int, res *[]int) {
	if n == 0 {
		*res = append(*res, num)
		return
	}
	digit := num % 10
	path := []int{digit - k}
	if k != 0 {
		path = append(path, digit+k)
	}
	for _, nextDigit := range path {
		if nextDigit >= 0 && nextDigit < 10 {
			dfs(n-1, k, num*10+nextDigit, res)
		}
	}
}

File: leetcode/src/00967_numbers_with_same_consecutive_differences/solution.rs
use std::collections::VecDeque;
use std::iter::FromIterator;

struct Solution;

impl Solution {
    pub fn nums_same_consec_diff(n: i32, k: i32) -> Vec<i32> {
        if n == 1 {
            return (0..10).collect();
        }
        let mut q = VecDeque::from_iter(1..10);
        for _ in 1..n {
            for _ in 0..q.len() {
                if let Some(front) = q.pop_front() {
                    let digit = front % 10;
                    let mut path = vec![digit - k];
                    if k != 0 {
                        path.push(digit + k);
                    }
                    for next_digit in path {
                        if (0..10).contains(&next_digit) {
                            q.push_back(front * 10 + next_digit);
                        }
                    }
                }
            }
        }
        q.into_iter().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let n = 3;
        let k = 7;
        assert_eq!(
            Solution::nums_same_consec_diff(n, k),
            vec![181, 292, 707, 818, 929]
        );

        let n = 2;
        let k = 1;
        assert_eq!(
            Solution::nums_same_consec_diff(n, k),
            vec![10, 12, 21, 23, 32, 34, 43, 45, 54, 56, 65, 67, 76, 78, 87, 89, 98]
        );

        let n = 1;
        let k = 1;
        assert_eq!(
            Solution::nums_same_consec_diff(n, k),
            (0..10).collect::<Vec<_>>()
        );
    }
}
